### PR TITLE
Add YAML support

### PR DESCRIPTION
**Pull request by mavam**
To be rebased onto master after #1017 is merged. Ignore everything up to and including 6b4e564.

### CODE REVIEW

File: libvast/src/data.cpp
Comment (dominiklohmann) [inline libvast/src/data.cpp:385]:
```suggestion
caf::expected<data> from_yaml(std::string str) {
  try {
    // Maybe one glory day in the future it will be possible to perform a
    // single pass over the input without incurring a copy.
    auto node = YAML::Load(std::move(str));
    return parse(std::move(node));
```

If a function must copy a parameter, it should take that parameter by value. Not a reference, not a view, but a value.

Imagine you have a `std::string` on the calling side that you don't need anymore after calling `from_yaml`. With the function taking a `std::string_view` and then copying it explicitly, you cannot avoid a copy. Furthermore, taking a `std::string_view` signals to the reader that the function does not copy its argument, which is misleading in this case.

**Review by dominiklohmann — CHANGES_REQUESTED** [review]

File: libvast/vast/concept/parseable/vast/yaml.hpp
Comment (dominiklohmann) [inline libvast/vast/concept/parseable/vast/yaml.hpp:27]: Shouldn't this be using `narrow_cast` over `narrow`?

File: libvast/src/data.cpp
Comment (dominiklohmann) [inline libvast/src/data.cpp:427]: It'd be more idiomatic to append `YAML::EndMap` in a scope guard that you set up immediately after appending `YAML::BeginMap`. The same applies to all other places.

File: libvast/src/data.cpp
Comment (dominiklohmann) [inline libvast/src/data.cpp:398]: Can we have this function do `static_assert(std::always_false_v<decltype(x)>, "unsupported type")` and instead list all types explicitly? Otherwise this becomes easy to overlook in future refactorings.

File: libvast/src/data.cpp
Comment (dominiklohmann) [inline libvast/src/data.cpp:375]: Wouldn't `die(...)` be more appropriate here?

**Review by dominiklohmann — CHANGES_REQUESTED** [review]
Second pass.

File: libvast/src/data.cpp
Comment (mavam) [reply libvast/src/data.cpp:427]: While safer, it is still more complicated to read. If you start with the tutorial of the official package, you can immediately recognize the patttern.

If the control flow would be more convoluted, I'd fully agree, but adding RAII to these few lines are not adding much value IMO.

File: libvast/src/data.cpp
Comment (mavam) [reply libvast/src/data.cpp:398]: I'll provide overloads for all types and ditch `auto`. Then compiler will complain if an overload is missing.

File: libvast/src/data.cpp
Comment (mavam) [reply libvast/src/data.cpp:375]: Good call.

File: libvast/src/data.cpp
Comment (dominiklohmann) [reply libvast/src/data.cpp:427]: RAII isn't a matter of readability, but rather of security. yaml-cpp uses exceptions, and if the code between the begin and end statements throws, we still want the end statement to be executed.

File: libvast/src/data.cpp
Comment (mavam) [reply libvast/src/data.cpp:386]: > If a function must copy a parameter, it should take that parameter by value. Not a reference, not a view, but a value.

The function signature in yaml-cpp looks like this:

```cpp
YAML_CPP_API Node Load(const std::string& input);
```

So moving in there doesn't strike me as an improvement.

My bigger point was though: there could be 10 copies happening internally, it doesn't matter because it's not going to be measurable. I don't want to change the API when an internal yaml-cpp optimization gets rolled out in the future. Therefore, `std::string_view` is still providing us with most flexibility at the call site.

Comment (dominiklohmann) [reply libvast/src/data.cpp:386]: > Therefore, `std::string_view` is still providing us with most flexibility at the call site.

There is no flexibility here at all. The function is taking a string view only to create a new string from it. Let's make this clearer on our end regardless of the yaml-cpp API: Take a string when you want a string, take a string view when you want a string view. Here you want a string, not a string view. If I read this code a year down the road I might be confused as to why a copy is happening in a function that takes a string view.

File: libvast/src/data.cpp
Comment (dominiklohmann) [reply libvast/src/data.cpp:399]: Good call.

File: libvast/src/data.cpp
Comment (mavam) [reply libvast/src/data.cpp:386]: It looks like I'm not getting my point across very well, sorry for that. Let's ignore the implementation for a moment (because it doesn't matter performance-wise). In your opinion, what string type should a `to_yaml` function take as argument for maximum flexibility and modern practices?

Comment (dominiklohmann) [reply libvast/src/data.cpp:386]: It should take a `std::string` if its implementation is required to copy and/or modify its input.

It should take a `std::string_view` if no copying occurs, ever.

Comment (mavam) [reply libvast/src/data.cpp:386]: I think this is where we have a different approach: I wouldn't sacrifice API architecture because of an implementation detail *unless* it's on the critical path.

Comment (dominiklohmann) [reply libvast/src/data.cpp:386]: Let's have @lava be the decider.

File: libvast/src/data.cpp
Comment (dominiklohmann) [reply libvast/src/data.cpp:435]: @lava mind being the decider here, again?

Comment (lava) [reply libvast/src/data.cpp:437]: Several issues with this:
 - If yaml-cpp indeed uses exceptions on the print path, the `to_yaml()` function below needs a try-catch block as well
 - If printing can throw, we can't blindly print in a scope guard since that might throw again. Instead we'd have to check `std::active_exceptions()`.
 - Printing `EndMap` is part of the happy path only. There is no guarantee that `Yaml::EndMap` is syntactically correct when an exception is thrown (i.e. if the exception happens between printing a key and a value). So we have to discard the result anyways after an exception.
- We don't leak resources by not printing `YAML::EndMap`

So, I'm in favor of leaving the code as-is. The first bullet point should still be addressed, though.

File: libvast/src/data.cpp
Comment (lava) [reply libvast/src/data.cpp:388]: I'd say let's not be too dogmatic for our own good:

```
caf::expected<data> from_yaml(const std::string& str)
```

* Never does a copy
* Makes sense from an API perspective
* Is actually the type that we want in the implementation

File: libvast/src/data.cpp
Comment (lava) [inline libvast/src/data.cpp:378]: That's pretty harsh, why not just throw `YAML::Exception` and fail to parse the file?

Comment (mavam) [reply libvast/src/data.cpp:378]: The idea is that this is more like a developer issue. We enumerated all switch statements above, so we should never reach this code. If so, it's a logic error. In order to make the compiler happy (and not spit out a warning), I plugged in a `[[noreturn]]` function for which we conveniently have `die`.

**Comment by mavam** [comment]
> If yaml-cpp indeed uses exceptions on the print path, the to_yaml() function below needs a try-catch block as well

@lava I could not find a single `throw` in the emitter code in yaml-cpp, that's why I didn't wrap it into a `try`-`catch` block.

File: libvast/src/data.cpp
Comment (mavam) [reply libvast/src/data.cpp:388]: >  if I have to choose between the two options, I'd go with keeping `string_view`

Okay, then I won't make any further API changes.

File: libvast/src/data.cpp
Comment (lava) [reply libvast/src/data.cpp:378]: I'm fine with leaving it as-is here, but in general we should be careful with jumping from "programmer error" to "kill the database", especially when there's a perfectly reasonable recovery path.

Related rant: https://lkml.org/lkml/2016/10/4/1

**Comment by mavam** [comment]
I've added a changelog as well. Before the next release, we may consolidate it with a single entry about the new config file format.

**Review by dominiklohmann — CHANGES_REQUESTED** [review]
Please install `libyaml-cpp-dev` in the bundled Dockerfile as well.

File: CMakeLists.txt
Comment (dominiklohmann) [inline CMakeLists.txt:865]:
```suggestion
```

Looks like a leftover.

**Comment by mavam** [comment]
> Please install `libyaml-cpp-dev` in the bundled Dockerfile as well.

Why didn't CI complain about this missing dependency, @dominiklohmann?

File: libvast/vast/concept/parseable/vast/yaml.hpp
Comment (dominiklohmann) [inline libvast/vast/concept/parseable/vast/yaml.hpp:42]:
```suggestion

} // namespace vast
```

File: libvast/vast/data.hpp
Comment (dominiklohmann) [inline libvast/vast/data.hpp:320]:
```suggestion
/// @returns The parsed YAML as data on success, or an error.
```

File: libvast/vast/data.hpp
Comment (dominiklohmann) [inline libvast/vast/data.hpp:325]:
```suggestion
/// @returns The YAML representation of x on success, or an error.
```

If you want to use highlighting, the proper way to do it for Doxygen / standardese would be `\p x` if I recall correctly. However, since we're not doing this anywhere else, let's just not do it here either.

**Review by dominiklohmann — APPROVED** [review]
Just cosmetic requests now. Approving already.

Please fix these and wait for CI before merging, especially because of the Docker CI.

**Comment by mavam** [comment]
@dominiklohmann I pushed a minor fixup and now the Docker build failed with this error:

```
Warning: apt-key output should not be parsed (stdout is not a terminal)
Executing: /tmp/apt-key-gpghome.rS81EKvBWf/gpg.1.sh --keyserver hkp://pool.sks-keyservers.net:80 --recv-keys 5C808C2B65558117
gpg: keyserver receive failed: Cannot assign requested address
```

Any objections if I merge nonetheless?

**Comment by mavam** [comment]
I got your approval in person, @dominiklohmann. 🙂

**Pull request merged**
